### ISSUE TITLE
bad memory reference while decompressing fuzzed input

### ISSUE THREAD

**Issue by jreiser** [upx/upx#862]
This issue tracker is **ONLY** used for reporting bugs.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?

Valgrind detected an out-of-bounds Read from memory when testing (`upx -t`) a fuzzed input.
## What should have happened?

Memory references should always be in bounds.
## How can we reproduce the issue?

1. build upx from source git-baee96 (tip of 'devel' branch on Wed Dec 11, 2024)
2. `valgrind upx/build/debug/upx -t clusterfuzz-testcase-minimized-test_packed_file_fuzzer-5090558997495808`
3.
```
                      Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2024
UPX git-baee96  Markus Oberhumer, Laszlo Molnar & John Reiser   Aug 19th 2024

testing clusterfuzz-testcase-minimized-test_packed_file_fuzzer-5090558997495808
==612622== Invalid read of size 1
==612622==    at 0x761FA6: ucl_nrv2e_decompress_safe_8 (n2e_d.c:76)
==612622==    by 0x5067D4: upx_ucl_decompress(unsigned char const*, unsigned int, unsigned char*, unsigned int*, int, upx_compress_result_t const*) (compress_ucl.cpp:177)
==612622==    by 0x4EF5E7: upx_decompress(unsigned char const*, unsigned int, unsigned char*, unsigned int*, int, upx_compress_result_t const*) (compress.cpp:161)
==612622==    by 0x6D52D1: ph_decompress(PackHeader&, XSpan::PtrOrSpan<unsigned char const>, XSpan::PtrOrSpan<unsigned char>, bool, Filter*) (packhead.cpp:335)
==612622==    by 0x6C312A: Packer::decompress(XSpan::PtrOrSpan<unsigned char const>, XSpan::PtrOrSpan<unsigned char>, bool, Filter*) (packer.cpp:306)
==612622==    by 0x670231: PackUnix::unpack(OutputFile*) (p_unix.cpp:686)
==612622==    by 0x6C117C: Packer::test() (packer.cpp:127)
==612622==    by 0x6C0DD0: Packer::doTest() (packer.cpp:107)
==612622==    by 0x6DE5D8: PackMaster::test() (packmast.cpp:272)
==612622==    by 0x757246: do_one_file(char const*, char*) (work.cpp:338)
==612622==    by 0x7579B3: do_files(int, int, char**) (work.cpp:424)
==612622==    by 0x53EEB5: upx_main(int, char**) (main.cpp:1321)
==612622==  Address 0x577be4f is 0 bytes after a block of size 329,727 alloc'd
==612622==    at 0x484280F: malloc (vg_replace_malloc.c:442)
==612622==    by 0x735CA9: MemBuffer::alloc(unsigned long long) (membuffer.cpp:180)
==612622==    by 0x66F9BF: PackUnix::unpack(OutputFile*) (p_unix.cpp:641)
==612622==    by 0x6C117C: Packer::test() (packer.cpp:127)
==612622==    by 0x6C0DD0: Packer::doTest() (packer.cpp:107)
==612622==    by 0x6DE5D8: PackMaster::test() (packmast.cpp:272)
==612622==    by 0x757246: do_one_file(char const*, char*) (work.cpp:338)
==612622==    by 0x7579B3: do_files(int, int, char**) (work.cpp:424)
==612622==    by 0x53EEB5: upx_main(int, char**) (main.cpp:1321)
==612622==    by 0x53EFB5: main (main.cpp:1385)
==612622== 
```
4.

## Please tell us details about your environment.

* UPX version used (`upx --version`): git-baee96
* Host Operating System and version: Fedora Linux 38
* Host CPU architecture: amd64
* Target Operating System and version: Fedora Linux 38
* Target CPU architecture: amd64

 Detailed original oss-fuzz report at https://oss-fuzz.com/testcase-detail/5090558997495808
 Original reproducible testcase at https://oss-fuzz.com/viewer?testcase_id=5090558997495808&key=815a6c1b-9b75-4a9f-ad6a-c2b160c3ef92  attached here as [5090558997495808.gz](https://github.com/user-attachments/files/18118424/5090558997495808.gz)  307 bytes, expands to 1730 bytes.

**Comment by gogo2464** [comment]
Could a malicious user input values to the program while running the program uncompressing with upx please?

**Comment by jreiser** [comment]
@gogo2464  UPX does not use file locking. Any user (program) that has Write permission on a file can write the file at any time, including while the file is open for reading by another program such as UPX.

**Comment by gogo2464** [comment]
Hi @jreiser thanks for the reply.

I stay confuse. Is the memery corruption exploitable in any way and what is the impact please?

**Comment by jreiser** [comment]
Yes, anyone who can write the file can corrupt the processing, although the access pattern of UPX during compression does reduce somewhat the exploitable windows of time.  [Edit:] This is one reason why the developer who uses UPX compression should immediately de-compress and compare with the original.

**Comment by gogo2464** [comment]
Very interesting! I know a website that uses upx to uncompress files online as a service. Sounds higtly potentially vulnerable.

**Comment by jreiser** [comment]
The complaint from valgrind is a Read of 1 byte beyond an allocated block.  The message from `upx -t` (or `upx -d`) is `Exception: compressed data violation`.   Of course it would be better if upx did not access that byte, but upx did detect derelict input, both with a message and a non-zero exit code (i.e., Failure).  Even if the bad Read by upx results in SIGSEGV, then that still implies "the iinput to upx is not good". So there is no security vulnerability.

**Comment by gogo2464** [comment]
in your opinion, might there be a chance that a loop occurs around the vulns that could reach an off by one vuln please?

**Comment by jreiser** [comment]
@gogo2464 In my opinion the valgrind complaint occurs because the `compressed data violation` is detected in a [logical] pipeline when the previous stage in the pipeline has already fetched the next byte.  This is a read-ahead violation of limited scope (1 byte).  There is no danger.
